R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+    dep})
+  dependencies <- lapply(dependencies, replace_with_lock, lock)
+  list(package = package, dependencies = dependencies)
+}
+
+strip_available_dependencies <- function(dependencies) {
+  dependencies <- lapply(dependencies
+    , function(package) {
+      if(package$name %in% pesky_namespaces) NULL
+      else package
+    })
+ dependencies[!vapply(dependencies, is.null, logical(1))]
+}
+
+#' We don't trust package authors to only put a package in once
+strip_duplicate_dependencies <- function(dependencies) {
+  dependencies[!duplicated(vapply(dependencies, function(d) d$name
+    , character(1)))]
+}
+
+#' Get the dependencies for a given package
+get_remote_dependencies <- function(package) {
+  UseMethod("get_remote_dependencies")
+}
+
+#' If a package is local we just read from the directory given
+get_remote_dependencies.local <- function(package) {
+  description_name <- paste0(package$dir, .Platform$file.sep, "DESCRIPTION")
+  list(package = package
+    , dependencies =
+      dependencies_from_description(package, read.dcf(description_name)))
+}
+
+#' For packages on CRAN we will extract to a temporary directory when we
+#' download the accurate remote DESCRIPTION file. Because these are tarballs
+#' there is no simple way to extract only our desired file like we can with
+#' zipfiles using the unz function.
+get_remote_dependencies.CRAN <- function(package) {
+  original_version <- package$version
+  if (package$is_dependency_package) {
+    package$version <- NA
+  }
+  filepath <- download_package(package)
+
+  sep <- .Platform$file.sep
+  split_fp <- strsplit(filepath,sep)[[1]]
+  dirpath <- dirname(filepath)
+  file_list <- untar(filepath, list = TRUE)
+  description_name <- file_list[grepl(paste0("^[^", sep, "]+", sep
+    ,"DESCRIPTION$"), file_list)]
+  output <- untar(filepath, description_name, exdir = dirpath)
+  description_path <- paste0(dirpath, sep, description_name)
+  package$download_path <- filepath
+  package$version <- original_version
+  dcf <- read.dcf(file = description_path)
+  package$latest_version <- version_from_description(package$name, dcf)
+  unlink(description_path)
+  list(package = package, dependencies = dependencies_from_description(package, dcf))
+}
+
+#' Download the accurate remote DESCRIPTION file for a github repo.
+get_remote_dependencies.github <- function(package) {
+  output <- download_description_github(package)
+  list(package = output$package
+    , dependencies = dependencies_from_description(package, output$dcf))
+}
+
+download_description_github <- function(package) {
+  remote <- package$remote
+  filepath <- download_package(structure(
+    package,
+    class = c(remote, class(package))))
+  package$download_path <- filepath
+  file_list <- unzip(filepath, list = TRUE)
+  sep <- .Platform$file.sep
+  subdir <- ""
+  if (!is.null(package$subdir)){
+    subdir <- paste0(sep, package$subdir)
+  }
+  description_name <- file_list$Name[grepl(paste0("^[^", sep, "]+"
```

Comment:
You've already included `sep` in `subdir` above and are using `+` matching. This means you'll be looking for things like `//foo/DESCRIPTION` -- how is that possibly going to match? In general, I would prefer something like:

```r
file_list <- Filter(function(x) substring(x, 1, nchar(subdir)) == subdir, file_list)
description_name <- Find(function(x) basename(x) == "DESCRIPTION", file_list)
```